libs/utils/android/workloads/viewer.py

```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2015, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import re
+import os
+import logging
+
+from time import sleep
+
+from android import Screen, System
+from android.workload import Workload
+
+
+class Viewer(Workload):
+    """
+    Android generic Viewer workload
+    """
+    
+    # Let the system pick the best package
+    package = ''
+
+    def __init__(self, test_env):
+        super(Viewer, self).__init__(test_env)
+        self._log = logging.getLogger('Viewer')
+        self._log.debug('Workload created')
+
+        # Set of output data reported by the browser
+        self.db_file = None
+
+    def run(self, out_dir, uri, workload, collect=''):
+        """
+        Run viewer workload
+
+        :param out_dir: Path to experiment directory where to store results.
+        :type out_dir: str
+
+        :param uri: The URI of the item to display
+        :type location_search: str
+
+        :param workload: Function pointer to the workload to run once item
+        is opened and being viewed.
+        :type workload: func
```

Comment:
It does not seems so much more code than what you have in your Notebook.

I think we should be able to defined the class in the notebook itself, did not tried in a Notebook but something like that should work:

```python
class GMapsSwipe(Viewer):

    def __init__(self, test_env):
        super(GMapsSwipe, self).__init__(test_env)
        # Create script
        self._script = TargetScript(self._te, "gmaps_swiper.sh")
        # Define commands to execute during experiment
        for i in range(2):
            System.hswipe(self._script, 40, 60, 100, False)
            script.append('sleep 1')
            System.vswipe(self._script, 40, 60, 100, True)
            script.append('sleep 1')
            System.hswipe(self._script, 40, 60, 100, True)
            script.append('sleep 1')
            System.vswipe(self._script, 40, 60, 100, False)
            script.append('sleep 1')
        self._script.push()

    def workload(self):
        self._script.run()

def experiment():
    # Configure governor
    target.cpufreq.set_all_governors('sched')

    # Create the workload and run it
    wload = GMapsSwipe()
    wload.run(out_dir=te.res_dir,
        collect="ftrace",
        uri="https://goo.gl/maps/eAbA8ZxPAy32")
        
    # Dump platform descriptor
    te.platform_dump(te.res_dir)
```

Isn't it?

In that case I would say we have pretty much the same flexibility than using a function pointer... but we also have a structure which allows to easily deploy the new class under workloads and merge it in LISA.